xbmc/guilib/GUIFontTTFGLBase.cpp

```diff
@@ -0,0 +1,256 @@
+
+#include "GUIFont.h"
+#include "GUIFontManager.h"
+#include "ServiceBroker.h"
+#include "Texture.h"
+#include "TextureManager.h"
+#include "gui3d.h"
+#include "rendering/MatrixGL.h"
+#include "utils/GLUtils.h"
+#include "utils/log.h"
+#include "windowing/GraphicContext.h"
+
+#include <cassert>
+
+// stuff for freetype
+#include <ft2build.h>
+#include FT_FREETYPE_H
+#include FT_GLYPH_H
+#include FT_OUTLINE_H
+
+namespace
+{
+constexpr int ELEMENT_ARRAY_MAX_CHAR_INDEX{1000};
+}
+
+CGUIFontTTFGLBase::CGUIFontTTFGLBase(const std::string& strFileName) : CGUIFontTTF(strFileName)
+{
+  m_updateY1 = 0;
+  m_updateY2 = 0;
+  m_textureStatus = TEXTURE_VOID;
+}
+
+CGUIFontTTFGLBase::~CGUIFontTTFGLBase()
+{
+  // It's important that all the CGUIFontCacheEntry objects are
+  // destructed before the CGUIFontTTFGLBase goes out of scope, because
+  // our virtual methods won't be accessible after this point
+  m_dynamicCache.Flush();
+  DeleteHardwareTexture();
+}
+
+CVertexBuffer CGUIFontTTFGLBase::CreateVertexBuffer(const std::vector<SVertex>& vertices) const
+{
+  assert(vertices.size() % 4 == 0);
+  GLuint bufferHandle = 0;
+
+  // Do not create empty buffers, leave buffer as 0, it will be ignored in drawing stage
+  if (!vertices.empty())
+  {
+    // Generate a unique buffer object name and put it in bufferHandle
+    glGenBuffers(1, &bufferHandle);
+    // Bind the buffer to the OpenGL context's GL_ARRAY_BUFFER binding point
+    glBindBuffer(GL_ARRAY_BUFFER, bufferHandle);
+    // Create a data store for the buffer object bound to the GL_ARRAY_BUFFER
+    // binding point (i.e. our buffer object) and initialise it from the
+    // specified client-side pointer
+    glBufferData(GL_ARRAY_BUFFER, vertices.size() * sizeof(SVertex), vertices.data(),
+                 GL_STATIC_DRAW);
+    // Unbind GL_ARRAY_BUFFER
+    glBindBuffer(GL_ARRAY_BUFFER, 0);
+  }
+
+  return CVertexBuffer(bufferHandle, vertices.size() / 4, this);
+}
+
+void CGUIFontTTFGLBase::DestroyVertexBuffer(CVertexBuffer& buffer) const
+{
+  if (buffer.bufferHandle != 0)
+  {
+    // Release the buffer name for reuse
+    glDeleteBuffers(1, (GLuint*)&buffer.bufferHandle);
+    buffer.bufferHandle = 0;
+  }
+}
+
+CTexture* CGUIFontTTFGLBase::ReallocTexture(unsigned int& newHeight)
+{
+  newHeight = CTexture::PadPow2(newHeight);
+
+  CTexture* newTexture = CTexture::GetTexture(m_textureWidth, newHeight, XB_FMT_A8);
+
+  if (!newTexture || newTexture->GetPixels() == NULL)
+  {
+    CLog::Log(LOGERROR,
+              "GUIFontTTFGL::CacheCharacter: Error creating new cache texture for size %f",
+              m_height);
+    delete newTexture;
+    return NULL;
+  }
+  m_textureHeight = newTexture->GetHeight();
+  m_textureScaleY = 1.0f / m_textureHeight;
+  m_textureWidth = newTexture->GetWidth();
+  m_textureScaleX = 1.0f / m_textureWidth;
+  if (m_textureHeight < newHeight)
+    CLog::Log(LOGWARNING, "%s: allocated new texture with height of %d, requested %d", __FUNCTION__,
```

Comment:
Please update to the current log standards using `{}`.

For more information please see the [libfmt documentation](https://fmt.dev/latest/syntax.html)